File: problem1.cpp
// Time Complexity : O(m*n)
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No


class Solution {
public:
    void dfs(vector<vector<int>>& image,int i,int j,int color,int paint)
    {
        if(i<0 || j<0)
            return;
        if(i>=image.size() || j>=image[0].size())
            return;
        if(image[i][j]!=color)
        {
            return;
        }
        image[i][j]=paint;
        dfs(image,i+1,j,color,paint);
        dfs(image,i-1,j,color,paint);
        dfs(image,i,j+1,color,paint);
        dfs(image,i,j-1,color,paint);
    }
    vector<vector<int>> floodFill(vector<vector<int>>& image, int sr, int sc, int newColor) {
        if(image.size() && image[sr][sc]!=newColor)
        {
            dfs(image,sr,sc,image[sr][sc],newColor);
        }
        return image;
    }
};

File: Problem1.cs
public class FloodFills
    {
        // Time Complexity : O(m * n) - have to traverse entire matrix
        // Space Complexity : O(m * n) - queue will have almost all elements from matrix in the worst case 
        // Did this code successfully run on Leetcode : Yes
        // Any problem you faced while coding this : No
        int[][] directions;

        public int[][] FloodFill(int[][] image, int sr, int sc, int color)
        {
            int m = image.Length;
            int n = image[0].Length;

            if (image == null || image.Length == 0) return image;
            int originalColor = image[sr][sc];

            if (image[sr][sc] == color)
                return image;

          
            directions = new int[][] {
                new int[] { 0, 1 }, //right
                new int[] { 0, -1 }, //left
                new int[] { 1, 0 }, //down
                new int[] { -1, 0 } //up
            };

            dfs(image, sr, sc, color, m, n, originalColor);

            return image;

        }

        public void dfs(int[][] image, int sr, int sc, int color, int m, int n, int originalColor)
        {
            //base case, 
            if(sr < 0 || sc < 0 || sr == m || sc == n || image[sr][sc] != originalColor)
            {
                return;
            }

            //logic
            image[sr][sc] = color;
            foreach (var dirs in directions)
            {
                int nr = sr + dirs[0];
                int nc = sc + dirs[1];
                dfs(image, nr, nc, color, m, n, originalColor);
            }

        }
    }
